댓글 작성 로직 및 게시글 상세조회 시 응답에 댓글들을 같이 조회하도록 수정 (#WA-212)

Donggwon-Seo opened the pull request:
# What is this PR?🔍
- Jira Issues: WA-212

</br>

## Changes📝
- 댓글 작성 관련 로직 및 테스트 코드 구현
- 기존 게시글 조회 시 QueryDsl을 사용하여 조회 로직을 작성하여서 댓글 조회 로직을 추가
- 1:N 관계여서 한 번에 게시글 조회 로직을 하는 방향은 복잡하여 게시글 조회와 댓글 조회 로직을 분리
- 댓글이 없는 경우 조회 로직에서 null을 반환하는데 DTO에서 빈 리스트를 반환하도록 생성자 정의
- 댓글 작성자와 게시글 작성자가 다를 경우 false 같은경우 true를 응답해주도록 구현
- 댓글이 있는 경우 게시글 조회 관련 ControllerTest 코드 추가
-  @getter 어노테이션 삭제 및 Getter 메소드 추가
- 댓글 작성시 댓글 내용이 없는 것을 검증하는 어노테이션을 @notblank로 변경

Donggwon-Seo committed:
feat : 댓글 작성 로직 구현

- 댓글 작성은 회원만 가능하도록 구현
- 요청 응답 시 댓글 id와 게시글 id만 넘기도록 구현

Donggwon-Seo committed:
test : 댓글 관련 테스트 코드 작성

- 컨트롤러, 서비스, 레포지토리, 통합 테스트 작성
- 레포지토리 테스트 작성 시 Comment 관련 엔티티들을 커스텀하기 위한 CommentCompositeCustomizer 작성
- 댓글 관련 엔터티를 자동으로 값을 넣어주도록 하기 위해 NotSaveCommentCustomization 작성
- 댓글 작성 요청시 Request DTO에 @Notnull 어노테이션이 빠진 부분 수정
- Board 엔터티 내에서 Comment 엔터티와의 양방향 연관 관계를 위한 메소드 추가
- 양방향 연관 관계를 위한 메소드의 접근 제한자 논의 필요

2dongyeop commented on src/main/java/io/wisoft/wasabi/domain/comment/Comment.java:
```diff
@@ -0,0 +1,51 @@
+import static jakarta.persistence.FetchType.LAZY;
+import static jakarta.persistence.GenerationType.IDENTITY;
+
+@Entity
+@Table(name = "comments")
+@Getter
+@Setter
+public class Comment extends BaseTimeEntity {
```
롬복을 최대한 안쓰기로 했던 것 같으니 필요한 getter만 직접 만들고, setter는 제거해도 될 것 같아요.

2dongyeop commented on src/main/java/io/wisoft/wasabi/domain/board/Board.java:
```diff
@@ -48,11 +51,19 @@ public class Board extends BaseTimeEntity {
         member.getBoards().add(this);
     }
 
+    public void addComment(final Comment comment) {
+        comments.add(comment);
+        comment.setBoard(this);
+    }
+
```
비즈니스 로직에서는 사용되지 않고, 테스트 코드에서만 사용되는 것 같습니다.

또한 setter 생성시 위에 작성한 `setMember()` 와 같이 private으로 연관관계 편의 메서드를 만드는게 좋아 보여요.

2dongyeop commented on src/main/java/io/wisoft/wasabi/domain/comment/dto/WriteCommentRequest.java:
```diff
@@ -0,0 +1,9 @@
+
+public record WriteCommentRequest(
+        @NotNull Long boardId,
+        @NotNull String content
```
`@NotBlank`로 수정할 수 있을 것 같아요.

Donggwon-Seo committed:
test : 댓글 관련 테스트 코드 수정

Donggwon-Seo committed:
feat : 게시글 상세 조회 시 댓글들을 포함하여 응답하도록 수정

- 기존 게시글 조회 시 QueryDsl을 사용하여 조회 로직을 작성하여서 댓글 조회 로직을 추가
- 1:N 관계여서 한 번에 게시글 조회 로직을 하는 방향은 복잡하여 게시글 조회와 댓글 조회 로직을 분리
- 댓글이 없는 경우 조회 로직에서 null을 반환하는데 DTO에서 빈 리스트를 반환하도록 생성자 정의
- 댓글 작성자와 게시글 작성자가 다를 경우 false 같은경우 true를 응답해주도록 구현
- 댓글이 있는 경우 게시글 조회 관련 ControllerTest 코드 추가
-  @Getter 어노테이션 삭제 및 Getter 메소드 추가
- 댓글 작성시 댓글 내용이 없는 것을 검증하는 어노테이션을 @NotBlank로 변경

Donggwon-Seo commented:
Merge 할 때 application.yml 의 dev는 빼야합니당

Donggwon-Seo committed:
Merge branch 'develop' into feature/WA-212

2dongyeop commented on src/main/java/io/wisoft/wasabi/domain/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+        this.boardRepository = boardRepository;
+    }
+
+    public WriteCommentResponse writeComment(final WriteCommentRequest request, final Long memberId) {
```
트랜잭션 처리(`@Transactional`)가 누락된 것 같아요. 클래스 레벨에 `@Transactional(readOnly=true)` 처리도 추가하는건 어떨까요?

2dongyeop commented on src/main/java/io/wisoft/wasabi/global/exception/GlobalExceptionHandler.java:
```diff
@@ -48,13 +48,13 @@ public ResponseEntity<Response<?>> handleBusinessException(final BusinessExcepti
-
-        return buildResponse(ResponseType.UNCAUGHT_ERROR);
-    }
+//    @ExceptionHandler(RuntimeException.class)
+//    public ResponseEntity<Response<?>> handleRuntimeException(final RuntimeException ex) {
+//
+//        logger.info("\n [Error] RuntimeException : ErrorMessage : {}  Path: {}", ex.getMessage(), ex.fillInStackTrace());
+//
+//        return buildResponse(ResponseType.UNCAUGHT_ERROR);
+//    }
```
요 주석 처리한 부분은 어떤 내용인가요??

Donggwon-Seo replied:
제가 예외처리 된 거 없앨라고 하다가 누락했네용 수정했습니당

2dongyeop commented:
> Merge 할 때 application.yml 의 dev는 빼야합니당

개발할 때 application.yml에 dev 추가해서 하지 마시고, IDE에서 런타임시 프로필 지정해서 개발하시는게 더 좋을 것 같아요.

Donggwon-Seo committed:
fix : 패키지 구조 변경에 따른 예외 발생 및 회원 가입 시 AOP가 적용되는 타겟 범위 수정

- 예외발생 시 domain 하위의 모든 패키지의 하위 application 패키지 하위의 Service가 포함된 범위로 수정
- 회원가입 시 domain 하위의 web 경로 추가
- @Transactional을 클래스 레벨에 추가
